ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -109,22 +165,54 @@ long getMaxBlobSize() {
    * for a point in time within the specified range.
    * The following data are considered as valid data for this API:
    * 1. PUT with no expiry and no corresponding DELETE
-   * 2. PUT expiring at t_exp but t_exp_ref < t_exp
-   * 3. PUT with corresponding DELETE at time t_delete but t_del_ref < t_delete
+   * 2. PUT expiring at t_exp but t_ref < t_exp
+   * 3. PUT with corresponding DELETE at time t_delete but t_ref < t_delete
    * 4. DELETE record
-   * For this API, t_del_ref is based on the given {@link TimeRange} and t_exp_ref is the time when the API is called.
+   * For this API, t_ref is specified by the given {@link TimeRange}.
    * @param timeRange the reference {@link TimeRange} at which the data is requested. Defines both the reference time
    *                  and the acceptable resolution.
    * @return a {@link Pair} whose first element is the time at which stats was collected (in ms) and whose second
    * element is the valid data size for each segment in the form of a {@link NavigableMap} of segment names to
    * valid data sizes.
+   * @throws StoreException
    */
   Pair<Long, NavigableMap<String, Long>> getValidDataSizeByLogSegment(TimeRange timeRange) throws StoreException {
-    long deleteReferenceTimeInMs = timeRange.getEndTimeInMs();
-    long expirationReferenceTimeInMs = time.milliseconds();
-    NavigableMap<String, Long> validSizePerLogSegment =
-        collectValidDataSizeByLogSegment(deleteReferenceTimeInMs, expirationReferenceTimeInMs);
-    return new Pair<>(deleteReferenceTimeInMs, validSizePerLogSegment);
+    if (!enabled.get()) {
+      throw new StoreException("BlobStoreStats is not enabled or closing", StoreErrorCodes.Store_Shutting_Down);
+    }
+    ScanResults currentScanResults = scanResults.get();
+    Pair<Long, NavigableMap<String, Long>> retValue = null;
+    long referenceTimeInMs = getLogSegmentRefTimeMs(currentScanResults, timeRange);
+    if (referenceTimeInMs != REF_TIME_OUT_OF_BOUNDS) {
+      retValue = currentScanResults.getValidSizePerLogSegment(referenceTimeInMs);
+    } else {
+      if (isScanning && getLogSegmentRefTimeMs(indexScanner.newScanResults, timeRange) != REF_TIME_OUT_OF_BOUNDS) {
+        try {
+          boolean waitSuccess;
+          scanLock.lock();
+          if (isScanning) {
+            waitSuccess = waitCondition.await(waitTimeoutInSecs, TimeUnit.SECONDS);
```

Comment:
I used a passed in config and set the default to 2 minutes. I thought a config makes sense because the wait time changes depending on the configuration of the stores (size, speed, etc.)